meta-client/src/test/java/org/apache/helix/metaclient/puppy/AbstractPuppy.java

```diff
@@ -0,0 +1,79 @@
+package org.apache.helix.metaclient.puppy;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.helix.metaclient.api.MetaClientInterface;
+
+import java.util.HashMap;
+
+/**
+ * AbstractPuppy object contains interfaces to implement puppy and main logics to manage puppy life cycle
+ */
+public abstract class AbstractPuppy implements Runnable {
+
+  protected MetaClientInterface<String> _metaclient;
+  protected PuppySpec _puppySpec;
+  public HashMap<String, Integer> _eventChangeCounterMap;
+  public int _unhandledErrorCounter;
```

Comment:
I'd suggest try using getters instead of direct access. Besides, I think protected should be enough scope.

Reply:
I agree with protected security but what I am not quite sure I understands the use case for a getter method here. I can create a helper method that increments the counter but it would also be accessing the member variable the same way.